File: 2k22/Practice Session 3/Sherlock and Watson Gym Secrets.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll M = 1e9+7;

ll modpow(ll base, ll pow, ll mod){
    if (pow==0) return 1%mod;
    if (pow==1)return base%mod;

    ll res = modpow(base*base%mod, pow/2, mod);
    if (pow%2==1) return (base*res)%mod;
    return res;
}
ll solve(){
    ll a, b, n, k;
    cin>>a>>b>>n>>k;
    vector<ll> cnts(k);
    for(int i =1; i<=min(n, k);i++){
        ll num = ((n-i)/k+1)%M;
        ll modp = modpow(i, b, k);
        cnts[modp] = (cnts[modp]+num)%M;
    }
    ll res = 0;
    for(int i =1; i<=min(n, k); i++) {
        ll num = ((n-i)/k+1)%M;
        ll targmod = (k-modpow(i, a, k))%k;
        res+=(cnts[targmod]*num)%M;
        if (modpow(i, b, k)==targmod)
            res+=M-num;
    }
    return res%M;
}
int main(){
    int T;
    cin>>T;
    for(int i = 1; i<=T; ++i){
        cout<<"Case #"<<i<<": "<<solve();
    }
}

File: 2k22/Practice Session 3/Sherlock and Watson Gym Secrets.py

def modpow(base, power, mod):
    if (power == 0): return 1%mod
    if (power == 1): return base%mod
    res = modpow(base*base%mod, power//2, mod)
    if (power%2==1): return (base*res)%mod
    return res

M=1000000007
def solve():
    A, B, N, K = map(int, input().split())
    cnts = [0]*K
    for i in range(1, min(N, K)+1):
        num = ((N-i)//K+1)%M
        modp = modpow(i, B, K)
        cnts[modp] = (cnts[modp]+num)%M
    
    res =0
    for i in range(1, min(N, K)+1):
        num = ((N-i)//K+1)%M
        targmod = (K-modpow(i, A, K))%K
        res+=(cnts[targmod]*num)%M
        
        if modpow(i, B, K) == targmod:
            res+= (M-num)
    return res%M

import time
start_time = time.time()
T = int(input())
for t in range(T):
    print(f"Case #{t+1}: {solve()}")
print("--- %s seconds ---" % (time.time() - start_time))